### PR TITLE
enhance: Ensure ImportV2 waits for the index to be built and refine some logic

### PR DESCRIPTION
**Pull request by bigsheeper**
Feature Introduced:
1. Ensure ImportV2 waits for the index to be built

Enhancements Introduced:
1. Utilization of local time for timeout ts instead of allocating ts from rootcoord.
3. Enhanced input file length check for binlog import.
4. Removal of duplicated manager in datanode.
5. Renaming of executor to scheduler in datanode.
6. Utilization of a thread pool in the scheduler in datanode.

issue: https://github.com/milvus-io/milvus/issues/28521

### CODE REVIEW

**Comment by bigsheeper** [comment]
rerun ut

**Review by xiaocai2333 — APPROVED** [review]

File: internal/datacoord/import_checker.go [hunk 1]
```diff
@@ -241,64 +241,74 @@ func (c *importChecker) checkPreImportingJob(job ImportJob) {
...
-			}
-		}
+	segmentIDs := lo.FlatMap(tasks, func(t ImportTask, _ int) []int64 {
+		return t.(*importTask).GetSegmentIDs()
+	})
+
+	// Verify completion of index building for imported segments.
+	unindexed := c.meta.indexMeta.GetUnindexedSegments(job.GetCollectionID(), segmentIDs)
+	for _, segmentID := range unindexed {
+		c.buildIndexCh <- segmentID // accelerate index building
```
[the remark this answers is not in the document]
Comment (xiaocai2333) [reply internal/datacoord/import_checker.go:260]: ignore it.

File: internal/datanode/services.go [hunk 2]
```diff
@@ -515,7 +515,7 @@ func (node *DataNode) DropImport(ctx context.Context, req *datapb.DropImportRequ
 		return merr.Status(err), nil
 	}
 
-	node.importManager.Remove(req.GetTaskID())
+	node.impTaskMgr.Remove(req.GetTaskID())
```
Comment (czs007) [inline internal/datanode/services.go:518]: rename to importTaskMgr

File: internal/datanode/data_node.go [hunk 3]
```diff
@@ -286,8 +287,8 @@ func (node *DataNode) Init() error {
 
 		node.writeBufferManager = writebuffer.NewManager(syncMgr)
 
-		node.importManager = importv2.NewManager(node.syncMgr, node.chunkManager)
-
+		node.impTaskMgr = importv2.NewTaskManager()
+		node.impScheduler = importv2.NewScheduler(node.impTaskMgr, node.syncMgr, node.chunkManager)
```
Comment (czs007) [inline internal/datanode/data_node.go:291]: rename to importScheduler

File: internal/datacoord/import_checker.go [hunk 4]
```diff
@@ -241,64 +241,74 @@ func (c *importChecker) checkPreImportingJob(job ImportJob) {
...
+	}
+	if Params.DataCoordCfg.WaitForIndex.GetAsBool() && len(unindexed) > 0 {
```
Comment (czs007) [inline internal/datacoord/import_checker.go:262]: 10 30 30 30
recalculate progress

File: internal/datacoord/import_checker.go [hunk 5]
```diff
@@ -241,64 +241,74 @@ func (c *importChecker) checkPreImportingJob(job ImportJob) {
...
+		if segment == nil {
+			log.Warn("cannot find segment, may be compacted", zap.Int64("segmentID", segmentID))
+			return false
+		}
+		return segment.GetIsImporting()
+	})
+
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 	err := c.sm.FlushImportSegments(ctx, job.GetCollectionID(), unfinished)
```
Comment (czs007) [inline internal/datacoord/import_checker.go:278]: move to scheduler

File: internal/datanode/importv2/scheduler.go [hunk 6]
```diff
@@ -79,39 +77,44 @@ func (e *executor) Start() {
...
+				switch task.GetType() {
+				case PreImportTaskType:
+					fs := s.PreImport(task)
+					futures[task.GetTaskID()] = fs
+				case ImportTaskType:
+					fs := s.Import(task)
+					futures[task.GetTaskID()] = fs
+				}
+			}
+			for taskID, fs := range futures {
```
Comment (czs007) [inline internal/datanode/importv2/scheduler.go:96]: The lifespan of the futures object here is excessively prolonged.

Comment (czs007) [reply internal/datanode/importv2/scheduler.go:96]: Add Done to Future

**Comment by bigsheeper** [comment]
rerun ut

File: internal/datanode/importv2/scheduler.go [hunk 7]
```diff
@@ -126,33 +131,42 @@ func WrapLogFields(task Task, fields ...zap.Field) []zap.Field {
 	return res
 }
 
-func (e *executor) handleErr(task Task, err error, msg string) {
+func tryFreeFutures(futures map[int64][]*conc.Future[any]) {
+	for k, fs := range futures {
+		fs = lo.Filter(fs, func(f *conc.Future[any], _ int) bool {
+			return !f.Done()
```
Comment (czs007) [inline internal/datanode/importv2/scheduler.go:137]: no need to call f.Await() for futures which is done ?

File: internal/datacoord/import_scheduler.go [hunk 8]
```diff
@@ -295,12 +300,14 @@ func (s *importScheduler) processInProgressImport(task ImportTask) {
...
-			op := UpdateBinlogsOperator(info.GetSegmentID(), info.GetBinlogs(), info.GetStatslogs(), nil)
-			err = s.meta.UpdateSegmentsInfo(op)
+			op1 := UpdateBinlogsOperator(info.GetSegmentID(), info.GetBinlogs(), info.GetStatslogs(), nil)
+			op2 := UpdateStatusOperator(info.GetSegmentID(), commonpb.SegmentState_Flushed)
+			err = s.meta.UpdateSegmentsInfo(op1, op2)
 			if err != nil {
 				log.Warn("update import segment binlogs failed", WrapTaskLog(task, zap.Error(err))...)
 				return
 			}
+			s.buildIndexCh <- info.GetSegmentID() // accelerate index building
```
Comment (czs007) [inline internal/datacoord/import_scheduler.go:310]: how about non-block write ?

File: internal/datanode/importv2/scheduler.go (hunk 7 above)
Comment (bigsheeper) [reply internal/datanode/importv2/scheduler.go]: done

File: internal/datacoord/import_scheduler.go (hunk 8 above)
Comment (bigsheeper) [reply internal/datacoord/import_scheduler.go]: done

**Comment by bigsheeper** [comment]
rerun ut

**Pull request merged**